templates/partner_campaign.html

```diff
@@ -0,0 +1,388 @@
+    #intro p { font-size: 18px; }
+    #partner-campaign-header .partner.logo { text-align: left; }
+    #partner-campaign-header .origin.logo { text-align: right; }
+
+    #screenshot1 img,
+    #screenshot2 img,
+    #screenshot2-desktop img { left: 75px; }
+
+    .qrcode-wrapper p {
+      font-family: Lato, sans-serif;
+      font-size: 14px;
+      line-height: 26px;
+      text-align: center;
+    }
+    .qrcode-wrapper img {
+      margin: 0 auto;
+      margin-top: 10px;
+    }
+    .qrcode-wrapper {
+      background: #fff;
+      color: #000;
+      padding: 5px;
+      margin: 0 auto;
+      width: 190px;
+      height: 190px;
+      border-radius: 20px;
+    }
+    .hidden-mobile {
+      display: inherit;
+    }
+    .hidden-desktop {
+      display: none;
+    }
+    .desktop-split {
+      display: block;
+      max-width: 1024px;
+      margin: 0 auto;
+    }
+    .float-desktop {
+      float: left;
+      width: 50%;
+      margin: 0;
+    }
+    .page-splitter {
+      width: 100%;
+      margin: 50px auto;
+    }
+  }
+</style>
+{% endblock %}
+
+{% block extra_scripts %}
+  <script src="/static/js/qrcode.min.js" type="text/javascript"></script>
+  <script>
+    const isIOS = navigator.platform && /iPad|iPhone|iPod/.test(navigator.platform)
+    const referalCode = 'op:{{ referal_code }}'
+    const me = `https://www.originprotocol.com/partner/{{ referal_code }}`
+
+    const qrOptions = {
+      text: me,
+      width: 140,
+      height: 140,
+      colorDark : "#000000",
+      colorLight : "#ffffff",
+      correctLevel : QRCode.CorrectLevel.H
+    }
+    const qrcode1 = new QRCode('qrcode1', qrOptions)
+    const qrcode2 = new QRCode('qrcode2', qrOptions)
+
+    /**
+     * Copy a string to the clipboard
+     */
+    function codeToClipboard(code) {
+      code = code || referalCode
+
+      // Use navigator.clipboard if available (cleaner)
+      if (navigator.clipboard) {
+        return navigator.clipboard.writeText(code)
+      }
+
+      // The hard way because it's a crap browser
+      return new Promise((resolve, reject) => {
+        // Create element
+        const tmp = document.createElement('span')
+        tmp.style.display = 'none'
+        tmp.textContent = code
+        document.body.appendChild(tmp)
+
+        // Selection
+        const sel = document.getSelection()
+        sel.removeAllRanges()
+
+        // Range
+        const rng = document.createRange()
+        rng.selectNodeContents(tmp)
+
+        // Add the range to the selection
+        sel.addRange(rng)
+        console.debug('selected: ', sel.getRangeAt(0).cloneContents())
+
+        // Copy the text in the range
+        const res = document.execCommand('copy')
+        document.body.removeChild(tmp)
+        if (!res) {
+          reject('Failed to execute copy of text')
+        }
+        resolve()
+      })
+    }
+
+    /**
+     * handle download button click and go to the appropriate download page
+     */
```

Comment:
If you want you can use async/await patern. Personally I like it because of the less boilerplate. With it this function would look something like this (also changed code style to ES6)
: 
```
const downloadApp = async () => {
  let url = 'https://play.google.com/store/apps/details?id=com.origincatcher'
  if (isIOS) {
    url = 'https://apps.apple.com/us/app/origin-marketplace/id1446091928'
  }
  
  try {
    await codeToClipboard(referalCode)
    const win = window.open(url)
    // If we got snagged by a popup blocker(firefox) just go direct
    if (!win) {
      window.location = url
    }
  } catch (err) {
    console.error('Problem copying code to clipboard')
    console.error(err)
    window.location = url
  }
  
  return false
}
```

Reply:
Does async/await(or even returning promises) work cleanly with HTML event handlers(e.g. `onclick`)?

Reply:
yup it works cleanly, since `async` typed function is under the hood just a Promise. And html events can accept promises.